### ISSUE TITLE
Ltac2 uconstr scope

### ISSUE THREAD

**Issue by JasonGross** [coq/coq#17524]
I guess I should open a new issue for "please give a notation scope for uconstr that internalizes as Constr.Unsafe-based construction"?

_Originally posted by @JasonGross in https://github.com/coq/coq/issues/17522#issuecomment-1520659422_

-------

When writing performant code, it's essential to be able to avoid unnecessary typechecking.  It's convenient to be able to write abbreviated constants.  I'd therefore like `uconstr:(...)` and `open_uconstr:(...)` which desugar into `Constr.Unsafe` constructions.  `uconstr:(...)` should error on holes/implicits, while `open_uconstr:(...)` should replace individual holes with `open_constr:(_)`.

It would be even nicer if `uconstr:(...)` could perform partial type inference at desugaring time: replace antiquotations with rigid evars, run type inference, replace the rigid evars with the antiquotations.

**Comment by SkySkimmer** [comment]
uconstr is certainly not the right name for this considering it's used by ltac1 for something else

**Comment by JasonGross** [comment]
Is it clear that Ltac1 uconstr is something else?  I guess there's a difference between "hole" and "evar" in Ltac1 uconstr, and presumably the internal representation is different?

**Comment by SkySkimmer** [comment]
Ltac1 uconstr is Ltac2 preterm with shared bindings (so if you have `let x := uconstr:(whatever) in constr:(use x)` it works same as if `x` was bound to a `constr` or a regular binder, unlike ltac2 where `$x` and the proposed https://github.com/coq/coq/pull/17359 use different syntax from real term binders).

So for instance if you do `uconstr:(Prop Prop)` if you ever try to turn it into a `constr` you get a type error. With unsafe constrs you would get an ill-typed term.

**Comment by ppedrot** [comment]
I don't think unsafe constr makes sense for anything that will go through the pretyper. It will check stuff anyways.

**Comment by JasonGross** [comment]
> I don't think unsafe constr makes sense for anything that will go through the pretyper. It will check stuff anyways.

Yeah, I want something that is typed as `constr` in Ltac2.  I'm perfectly happy with it literally desugaring to `Constr.Unsafe.*` calls during Ltac2 definition time, and don't want to have it go through the pretyper unless/until I explicitly invoke `Constr.Unsafe.check` (or is it `Constr.check`?) (is this the pretyper, or something else?)

**Comment by SkySkimmer** [comment]
Constr.Unsafe.check is typing
The pretyper is pretype : preterm -> constr

>I'm perfectly happy with it literally desugaring to Constr.Unsafe.* calls during Ltac2 definition time

There is no point discussing this sort of implementation details when we don't have a spec
